### DISCUSSION TITLE
Misskey 2024.10.2-alpha.2で内部・外部の画像が表示できない

### DISCUSSION THREAD

**Discussion by erisunaga** [misskey-dev/misskey#14936] [Q&A]
### 💡 Summary

2024年10月にインスタンスを立てました。
当初`13.13.2`で立て（参考サイトがそうだったからで、深い意味はないです）、
先月？から下記不調（きっかけとかは不明、立てた当初は問題なかったです）で、
`2024.10.2-alpha.2`にアップデートした後も自力で対処できそうになく、こちらにあげさせていただきます。

# できないこと

- ユーザーのアイコン表示
    -  `13.13.2`のときは、猫モードだと耳が透けて見える
    - `2024.10.2-alpha.2`にしたら、元画像のざっくりした色のキャンバスで表示（画像の表示はできていない）
        - 画像のURLを叩くと404エラー（R2のアドレス） 
- 投稿画像の参照
     - リモートユーザーが投稿した画像は見れたり見れなかったり（見れない方が多い） 
     - 自分でアップした画像も404
- 新規プロフィール画像のクリップ
    - ずっとぐるぐる読み込み中
    - クリップせずにアップすることはできる

### 🥰 Expected Behavior

全ての画像が問題なく読み込める

### 🤬 Actual Behavior

ほとんどの画像が表示できない

### 💻 Frontend Environment

```markdown
* Macbook Air M1
    * OS: Sonoma 14.7 
* Browser:
    * Safari :バージョン17.6 (19618.3.11.11.5)
    * Chrome :バージョン: 130.0.6723.117（Official Build）（arm64）
* Server URL: misskey.eri-sunaga.com
* Misskey:2024.10.2-alpha.2
```

### 🛰 Backend Environment (for server admin)

```markdown
- Docker Compose: v2.29.7-desktop.1
    - Misskey: 2024.10.2-alpha.2
    - cloudflare : 2024.11.0-arm64
        - Cloudflare Tunnel
    - redis:7-alpine
    - postgres: 15-alpine
    - meilisearch:v1.1.1
- Cloudflare R2
```

### Do you want to address this bug yourself?

- [ ] Yes, I will patch the bug myself and send a pull request

**Comment by fruitriin** [comment]
オブジェクトストレージかメディアプロクシの設定がおかしそう

**Comment by erisunaga** [comment]
@fruitriin 
さっそくありがとうございます。
メディアプロキシは設定していなくて、あるとすればオブジェクトストレージとして使っているCloudflare R2かなと…
ただ、`13.13.2`のときは使っていない状態で事象が出ていたので、R2を導入してしまったことが余計に事態をややこしくしているかもしれません。。。
いったんR2を解除して、もう少しいろいろ叩いてみます。

**Comment by erisunaga** [comment]
R2を解除して叩いてみました

- 解除後、TLの画像は読み込まれるようになったかも
    - グローバルTLで確認
    - ![スクリーンショット 2024-11-11 6 33 55](https://github.com/user-attachments/assets/5ff5941b-879e-438b-9330-9687e1963ebd) 
    - 接続前も、アバターが表示できないことに気を取られてTLが表示できていたかはよく覚えていません。。。
    - →R2の設定**も**悪いかも
- アバターは相変わらず表示できない
    - 自分のアバターはあれこれ試したので表示できている
    - リモートユーザーのアバターは表示できない
        - `<img height="120" width="120" class="xkJBF" src="https://misskey.eri-sunaga.com/proxy/avatar.webp?url=https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F1d35dedc-713d-4ea5-975f-a385c90240c4&amp;avatar=1" alt="" loading="eager" decoding="async" tabindex="-1" style="display: none;">`
        - ![スクリーンショット 2024-11-11 6 50 57](https://github.com/user-attachments/assets/8c6b5ce8-6173-44f1-ba64-23a59acfc2a7)

**Reply by erisunaga** [reply]
<img width="713" alt="スクリーンショット 2024-11-11 6 57 56" src="https://github.com/user-attachments/assets/a65f3f68-8a0f-4df6-b146-475f4a491303">

**Comment by erisunaga** [comment]
`13.13.2`で発生していた内容はこれに似ているかも

> 13.13.2のときは、猫モードだと耳が透けて見える

https://scrapbox.io/defaultcf/Misskey_13.3.2_にアップデートした時のログ

**Comment by erisunaga** [comment]
> # できないこと
> ユーザーのアイコン表示
> - 13.13.2のときは、猫モードだと耳が透けて見える
> - 2024.10.2-alpha.2にしたら、元画像のざっくりした色のキャンバスで表示（画像の表示はできていない）
>     - 画像のURLを叩くと404エラー（R2のアドレス）

●そもそもうちはおひとり様なので以下を無効にしました
`リモートのファイルをキャッシュする`

●コントロールパネルからリモートファイルのキャッシュを全削除

●`default.yml`の下記について、↑でキャッシュを無効にしたので、推奨されているとおり`proxyRemoteFiles: true`にしました
```
# Media Proxy
#mediaProxy: https://example.com/proxy

# Proxy remote files (default: false)
proxyRemoteFiles: true
```

●Cloudflareの設定を見直しました
- ライフサイクルが7日に設定されていたのを、なしに変更
    - そりゃ消えて404になるよな、と思ったけど7日経たずに消えていたので直接は関係ないかも
    - ドライブの容量が100MBで、それを超過した分が消えていたのかも（ドキュメント流し読みなので違うかも）
- BaseURLに`https://`がついていなかったのでつけた

---

Cloudflare側の設定はいい感じっぽいのですが、相変わらず一部リモートユーザーのアバターは表示できません。

表示できない場合
`https://misskey.eri-sunaga.com/proxy/avatar.webp?url=https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%******&avatar=1`

表示できる場合
`https://misskey.eri-sunaga.com/proxy/avatar.webp?url=https%3A%2F%2Fmedia.mstdn.jp%2F******.png&avatar=1
`

どうも、うちのfilesを参照してるとダメ（ファイルがない）で、キャッシュを無効にしてリモートを直に見に行くようになってからはいけるみたいです。filesの参照をやめさせるにはどうすればよいのか…

あと、カスタム絵文字も不調です。

<img width="233" alt="スクリーンショット 2024-11-16 8 40 35" src="https://github.com/user-attachments/assets/8cc72181-1e7c-44f7-a0db-24fa5ae60283">

左は直接アップしたやつ
右はリモートサーバーから輸入したやつ。

コントロールパネルでカスタム絵文字を参照すると？になってしまう

![スクリーンショット 2024-11-16 8 56 26](https://github.com/user-attachments/assets/99eddded-f6c2-4463-98e3-4135a147780e)

ここで参照しているファイルは下記
`https://misskey.eri-sunaga.com/proxy/emoji.webp?url=https%3A%2F%2Fpub-dd2fff42f08f4cea91a62d1a0d5155a6.r2.dev%2Ffile%2F******.png&emoji=1`

直接アップした絵文字は？だけど使える。
リモートから輸入したやつは使えない

**Comment by erisunaga** [comment]
↑で試していた輸入カスタム絵文字は、問題発生時点でインポートしていたものなので、別のはじめましての絵文字をインポートしたらうまく行ったかも

`https://misskey.eri-sunaga.com/proxy/emoji.webp?url=https%3A%2F%2Fmisskey-r2.eri-sunaga.com%2Ffiles%*****.png&emoji=1`

**Comment by eternal-flame-AD** [comment]
他の鯖主から凄く似っている問題聞いたので自分のコミット関連性あるかもしれません

サーバーURLを1xまたはMedia proxyを2x以上変えた場合はデータベースのURLは既にProxy使用しているかの判定が乱れてAvatar/Emoji URLをネストされて安全上アタックのかないかわからないので拒否したかもしれない。このコミットをrevertして一時的な修正できるけどは安全上推奨しません。　https://github.com/misskey-dev/misskey/commit/48d1539f3be895b7aa8ecdd6c581e47a55cc9264

対策はちょっと難しい。。　完全に修正しする方法はDBにローカルユーザーAvatarとEmojiのProxy URLに記入しない方がいいと思います。 このセキュリティコミットがなくても同じ根本的な原因で他の鯖主も廃止されたMedia ProxyがDBに残存してシャットダウンできませんの問題も聞いた。


他の鯖主のため書いたテンプレート：

このSQL Queryはデータを変更しません、問題あるアバターが統計でぎるだけ (nested count > 0 ->  :warning: )　。

```sql
with typed as (
     select *,
         case
             when "avatarUrl" ~ '%2Fproxy%2F' then 'nested'
             when "avatarUrl" ~ '^https://[PROXY/CDN/etcはこれで]' then 'proxied'
             when "avatarUrl" ~ '^https://mi\.yumechi\.jp' then 'direct'
             else 'unknown'
         end AS proxy_type
     FROM "user"
     WHERE "avatarUrl" IS NOT NULL
 )
 select count(*),
     proxy_type
 from typed
 group by proxy_type;
+--------+------------+
| count  | proxy_type |
|--------+------------|
| 181360 | proxied    |
+--------+------------+
SELECT 1
```

以下のQueryは自分も他の問題のために使ったけど危険 :warning: です。完全に理解した上で使ってください。

```sql
begin transaction read write;

CREATE TEMPORARY TABLE "user_avatar_update" (
    "id" CHARACTER VARYING(32) PRIMARY KEY NOT NULL,
    "updatedAvatarUrl" CHARACTER VARYING(512) NOT NULL
)

CREATE INDEX "user_avatar_update_url" ON "user_avatar_update" ("updatedAvatarUrl")

with typed as (
    select *, 
        case when "avatarUrl" ~ '%2Fproxy%2F' then 'nested'
        when "avatarUrl" ~ '^https://mproxy\.mi\.yumechi\.jp' then 'proxied'
        when "avatarUrl" ~ '^https://mi\.yumechi\.jp' then 'direct' 
        else 'unknown' end 
    AS proxy_type 
    FROM "user" WHERE "avatarUrl" IS NOT NULL)
INSERT INTO "user_avatar_update" ("id", "updatedAvatarUrl")
SELECT "id", [ 特定の状況に違います ]  AS "updatedAvatarUrl" FROM typed WHERE proxy_type = 'direct'

SAVEPOINT collected;

-- 結果を確認 (URLをコピペにてブラウザにテストして！！）

SELECT * FROM "user_avatar_update" LIMIT 10;

UPDATE "user" SET "avatarUrl" = "user_avatar_update"."updatedAvatarUrl" FROM "user_avatar_update" WHERE "user"."id" = "user_avatar_update"."id"

SAVEPOINT updated;

-- Commit 前に結果を確認

with typed as (
    select *,
        case
            when "avatarUrl" ~ '%2Fproxy%2F' then 'nested'
            when "avatarUrl" ~ '^https://mproxy\.mi\.yumechi\.jp' then 'proxied'
            when "avatarUrl" ~ '^https://mi\.yumechi\.jp' then 'direct'
            else 'unknown'
        end AS proxy_type
    FROM "user"
    WHERE "avatarUrl" IS NOT NULL
)
select count(*),
    proxy_type
from typed
group by proxy_type;

select * from "user" limit 10;

COMMIT;
```

**Reply by erisunaga** [reply]
```sql

-- ご教示いただいたSQLを実行してみました。

WITH typed AS (
    SELECT 
        *,
        CASE
            WHEN "avatarUrl" ~ '%2Fproxy%2F' THEN 'nested'
            WHEN "avatarUrl" ~ '^https://misskey-r2.eri-sunaga.com/files' THEN 'proxied'
            WHEN "avatarUrl" ~ '^https://misskey\.eri-sunaga\.com' THEN 'direct'
            ELSE 'unknown'
        END AS proxy_type
    FROM "user"
    WHERE "avatarUrl" IS NOT NULL
)
SELECT 
    COUNT(*) AS count,
    proxy_type
FROM typed
GROUP BY proxy_type;

-- 結果

 count | proxy_type 
-------+------------
 25165 | direct
(1 row)

-- directの内容を確認すべく、proxyに該当するavatarUrlのurlを抽出

SELECT 
    REGEXP_REPLACE("avatarUrl", '^https://misskey\.eri-sunaga\.com/proxy/avatar\.webp\?url=', '') AS extracted_text
FROM "user"
WHERE "avatarUrl" ~ '^https://misskey\.eri-sunaga\.com/proxy/avatar\.webp\?url=';

-- 結果
                                                                                                                                                                                extracted_text                                                                                                                                                                                 
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F1d5716e6-0de9-4f78-9b8e-8b53b297e4fd&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2Fac0ae6d2-b336-4f01-a385-706fdd6e082d&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F06574676-2ee4-45a8-afc0-b9be9b45e2f5&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F938be048-9668-42a3-8213-ba257127839b&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F4b0d435d-672f-4582-aa51-056ce3e5e7c6&avatar=1
 misskey-r2.eri-sunaga.com%2F%2Fa1c525ab-5f4b-4137-bd11-4bc0d7475a06.png&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F03271ea8-f2a3-4b17-8aa7-23a526846494&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F51822fa7-b799-44ee-be71-a4b4a067fb68&avatar=1
 https%3A%2F%2Fmisskey.eri-sunaga.com%2Ffiles%2F16c9651a-6221-44de-ad53-a1133117783f&avatar=1
 （省略）

-- https%3A%2F%2Fmisskey.eri-sunaga.comを除外します。

SELECT 
    REGEXP_REPLACE("avatarUrl", '^https://misskey\.eri-sunaga\.com/proxy/avatar\.webp\?url=', '') AS extracted_text
FROM 
    "user"
WHERE 
    "avatarUrl" ~ '^https://misskey\.eri-sunaga\.com/proxy/avatar\.webp\?url='
    AND REGEXP_REPLACE("avatarUrl", '^https://misskey\.eri-sunaga\.com/proxy/avatar\.webp\?url=', '') NOT LIKE '%https%3A%2F%2Fmisskey.eri-sunaga.com%';

-- 結果

                                                                                                                                                                                extracted_text                                                                                                                                                                                 
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 misskey-r2.eri-sunaga.com%2F%2Fa1c525ab-5f4b-4137-bd11-4bc0d7475a06.png&avatar=1
 https%3A%2F%2Frss-mstdn.studiofreesia.com%2Ffile%2Faccounts%2Favatars%2F110%2F887%2F911%2F251%2F917%2F122%2Foriginal%2Fff378175b41e613e.png&avatar=1
 misskey-r2.eri-sunaga.com%2F%2F7fa2984c-36cf-460f-9049-49883943ea3e.jpeg&avatar=1
 https%3A%2F%2Ffiles.nakn.jp%2Fnaakey%2Fimages%2F874c9e58-f54c-415e-8730-cb81b64b6f9c.png&avatar=1
 misskey-r2.eri-sunaga.com%2F%2Fec6b33fb-dbd4-4d3c-92c3-646dd68b2b4e.webp&avatar=1
 misskey-r2.eri-sunaga.com%2F%2F4b55ca78-8ef0-4f00-a5bc-2132a2bc5e93.png&avatar=1
 misskey-r2.eri-sunaga.com%2F%2F4003d0ef-54af-4a09-b5c5-1572cfb15135.png&avatar=1
 https%3A%2F%2Fpub-dd2fff42f08f4cea91a62d1a0d5155a6.r2.dev%2Ffile%2F026d4f08-7236-4b2b-bea9-b8e45a40b816.png&avatar=1
 https%3A%2F%2Fmedia.odakyu.app%2Faccounts%2Favatars%2F113%2F249%2F263%2F990%2F249%2F138%2Foriginal%2F7e4e03736fe3b155.jpeg&avatar=1
 https%3A%2F%2Ffiles.nakn.jp%2Fnaakey%2Fimages%2Fwebpublic-49b799ba-a62e-4747-9cad-b0fcc038d304.webp&avatar=1
 https%3A%2F%2Fpub-dd2fff42f08f4cea91a62d1a0d5155a6.r2.dev%2Ffile%2Fca55ad51-9d09-4ce4-91cf-913805ab6567.png&avatar=1
```

## とりあえずまとめると

オブジェクトストレージなし時代
途中からリモートファイルをキャッシュし始める（`files`にどんどんデータが入ってくる）
　↓
Cloudflare R2を設定（いろいろいじってURLが複数回変わっている）
`misskey-r2.eri-sunaga.com//~`にリモートファイルも含めて入ってくる
　↓
リモートファイルのキャッシュをやめた
オブジェクトストレージのBaseURLの設定が正しくなった（たぶん）
-  `misskey-r2.eri-sunaga.com//`
    - `https://`をつける
    - 末尾の`/`は二重になるのでいらない

## url=https%3A%2F%2Fmisskey.eri-sunaga.com〜を参照している画像について

> どうも、うちのfilesを参照してるとダメ（ファイルがない）で、キャッシュを無効にしてリモートを直に見に行くようになってからはいけるみたいです。filesの参照をやめさせるにはどうすればよいのか…

これですが、リモートファイルのキャッシュをやめて数日経ったからか、だいぶ改善されてきていそうです。
（キャッシュの期限切れで外を見に行った？）（そのあたりの実装がどうなっているかは未確認）

---

出勤しなきゃなので一旦ここまでで、今後の対応とかはまた追って考えます汗
